Nomad 1.7.0 does not honor env VAULT_TOKEN

### Nomad version
```
Nomad v1.7.0
BuildDate 2023-12-07T08:28:54Z
Revision e4150e9703f3be6ee2339f0e45ff0801186e022b
```
### Operating system and Environment details
N/A

### Issue
Vault 1.7.0 is not capable of reading the vault token from the env variable VAULT_TOKEN. This is a regression from 1.6.4 and not mentioned as a breaking change in the release, thus I'm assuming it's a bug.


### Reproduction steps
Simplified repro:
`docker run --rm -it -e NOMAD_SKIP_DOCKER_IMAGE_WARN="trustme" -e VAULT_TOKEN="legit" hashicorp/nomad:1.7.0 agent -server -data-dir=/tmp -vault-enabled`
vs.
`docker run --rm -it -e NOMAD_SKIP_DOCKER_IMAGE_WARN="trustme" -e VAULT_TOKEN="legit" hashicorp/nomad:1.6.4 agent -server -data-dir=/tmp -vault-enabled`

#### Expected Result
Server starts

#### Actual Result
On 1.7.0:
> ==> Error starting agent: server setup failed: Failed to setup Vault client: Vault token must be set
    2023-12-07T14:26:33.833Z [INFO]  nomad: shutting down server
    2023-12-07T14:26:33.833Z [ERROR] nomad: failed to setup Vault client: error="Vault token must be set"
    2023-12-07T14:26:33.833Z [ERROR] agent: error starting agent: error="server setup failed: Failed to setup Vault client: Vault token must be set"

### Job file (if appropriate)

<!--
If possible please post relevant logs in the issue.
`
Logs and other artifacts may also be sent to: nomad-oss-debug@hashicorp.com

Please link to your Github issue in the email and reference it in the subject
line:

> To: nomad-oss-debug@hashicorp.com
>
> Subject: GH-1234: Errors garbage collecting allocs

Emails sent to that address are readable by all HashiCorp employees but are
*not* publicly visible.
-->

### Nomad Server logs (if appropriate)

### Nomad Client logs (if appropriate)

--- a/.changelog/19349.txt
+++ b/.changelog/19349.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+cli: Fixed a bug that caused the `nomad agent` command to ignore the `VAULT_TOKEN` and `VAULT_NAMESPACE` environment variables
+```

--- a/command/agent/command.go
+++ b/command/agent/command.go
@@ -278,6 +278,10 @@ func (c *Command) readConfig() *Config {
 		return nil
 	}
 
+	// Read Vault configuration for the default cluster again after all
+	// configuration sources have been merged.
+	defaultVault = config.defaultVault()
+
 	// Check to see if we should read the Vault token from the environment
 	if defaultVault.Token == "" {
 		defaultVault.Token = os.Getenv("VAULT_TOKEN")

--- a/command/agent/command_test.go
+++ b/command/agent/command_test.go
@@ -6,6 +6,7 @@ package agent
 import (
 	"math"
 	"os"
+	"path"
 	"path/filepath"
 	"strings"
 	"testing"
@@ -13,6 +14,7 @@ import (
 	"github.com/hashicorp/nomad/ci"
 	"github.com/hashicorp/nomad/helper/pointer"
 	"github.com/mitchellh/cli"
+	"github.com/shoenig/test/must"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
@@ -501,6 +503,125 @@ func TestIsValidConfig(t *testing.T) {
 			assert.False(t, result)
 			require.Contains(t, mui.ErrorWriter.String(), tc.err)
 			t.Logf("%s returned: %s", tc.name, mui.ErrorWriter.String())
+		})
+	}
+}
+
+func TestCommand_readConfig(t *testing.T) {
+	// Don't run in parallel since this test modifies environment variables.
+
+	configFiles := map[string]string{
+		"base.hcl": `
+data_dir = "/tmp/nomad"
+region   = "global"
+
+server {
+  enabled = true
+}
+
+client {
+  enabled = true
+}
+`,
+		"vault.hcl": `
+data_dir = "/tmp/nomad"
+region   = "global"
+
+server {
+  enabled = true
+}
+
+client {
+  enabled = true
+}
+
+vault {
+  token     = "token-from-config"
+  namespace = "ns-from-config"
+}
+`,
+	}
+
+	configDir := t.TempDir()
+	for k, v := range configFiles {
+		err := os.WriteFile(path.Join(configDir, k), []byte(v), 0644)
+		must.NoError(t, err)
+	}
+
+	testCases := []struct {
+		name    string
+		args    []string
+		env     map[string]string
+		checkFn func(*testing.T, *Config)
+	}{
+		{
+			name: "vault token and namespace from env var",
+			args: []string{
+				"-config", path.Join(configDir, "base.hcl"),
+			},
+			env: map[string]string{
+				"VAULT_TOKEN":     "token-from-env",
+				"VAULT_NAMESPACE": "ns-from-env",
+			},
+			checkFn: func(t *testing.T, c *Config) {
+				must.Eq(t, "token-from-env", c.Vaults[0].Token)
+				must.Eq(t, "ns-from-env", c.Vaults[0].Namespace)
+			},
+		},
+		{
+			name: "vault token and namespace from config takes precedence over env var",
+			args: []string{
+				"-config", path.Join(configDir, "vault.hcl"),
+			},
+			env: map[string]string{
+				"VAULT_TOKEN":     "token-from-env",
+				"VAULT_NAMESPACE": "ns-from-env",
+			},
+			checkFn: func(t *testing.T, c *Config) {
+				must.Eq(t, "token-from-config", c.Vaults[0].Token)
+				must.Eq(t, "ns-from-config", c.Vaults[0].Namespace)
+			},
+		},
+		{
+			name: "vault token and namespace from flag takes precedence over env var and config",
+			args: []string{
+				"-config", path.Join(configDir, "vault.hcl"),
+				"-vault-token", "secret-from-cli",
+				"-vault-namespace", "ns-from-cli",
+			},
+			env: map[string]string{
+				"VAULT_TOKEN":     "secret-from-env",
+				"VAULT_NAMESPACE": "ns-from-env",
+			},
+			checkFn: func(t *testing.T, c *Config) {
+				must.Eq(t, "secret-from-cli", c.Vaults[0].Token)
+				must.Eq(t, "ns-from-cli", c.Vaults[0].Namespace)
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			ui := cli.NewMockUi()
+			defer func() {
+				// Print command stderr in case of a failed test case to help
+				// with debugging.
+				if t.Failed() {
+					t.Log(ui.ErrorWriter.String())
+				}
+			}()
+
+			cmd := &Command{
+				Ui:   ui,
+				args: tc.args,
+			}
+			for k, v := range tc.env {
+				t.Setenv(k, v)
+			}
+
+			got := cmd.readConfig()
+			must.NotNil(t, got)
+			tc.checkFn(t, got)
 		})
 	}
 }

